pyccel/codegen/printing/cwrappercode.py

```diff
@@ -56,1026 +50,697 @@ class CWrapperCodePrinter(CCodePrinter):
-    # Functions managing  the creation of wrapper body
-    # -------------------------------------------------------------------
+        body = IfSection(check, body)
+        return body
+        #TODO modify it to accepte multiple variables or list of variables ?
 
-    def _valued_variable_management(self, variable, collect_var, tmp_variable):
+    # --------------------------------------------------------------------
+    #                        Custom error generators
+    # --------------------------------------------------------------------
+    @staticmethod
+    def generate_datatype_error(variable):
         """
-        Responsible for creating the body collecting the default value of an valuedVariable
-        and the check needed.
-        if the valuedVariable is optional create body to collect the new value
-
+        Generate TypeError exception from the variable information (datatype, precision)
         Parameters:
         ----------
-        tmp_variable : Variable
-            The temporary variable  to hold result
-        variable     : Variable
-            The optional variable
-        collect_var  : Variable
-            variable which holds the value collected with PyArg_Parsetuple
+        variable : Variable
 
-        Returns
+        Returns:
         -------
-        section      :
-            IfSection
-        collect_body : List
-            list containing the lines necessary to collect the new optional variable value
+        func     : FunctionCall
+            call to PyErr_SetString with TypeError as exception and custom message
         """
+        dtype     = variable.dtype
 
-        valued_var_check  = PyccelEq(VariableAddress(collect_var), VariableAddress(Py_None))
-        collect_body      = []
-
-        if variable.is_optional:
-            collect_body  = [AliasAssign(variable, tmp_variable)]
-            section       = IfSection(valued_var_check, [AliasAssign(variable, Nil())])
-
+        if isinstance(dtype, NativeBool):
+            precision = ''
+        if isinstance(dtype, NativeComplex):
+            precision = '{} bit '.format(variable.precision * 2 * 8)
         else:
-            section       = IfSection(valued_var_check, [Assign(variable, variable.value)])
+            precision = '{} bit '.format(variable.precision * 8)
 
-        return section, collect_body
+        message = '"Argument must be {precision}{dtype}"'.format(
+                precision = precision,
+                dtype     = variable.dtype)
+        return PyErr_SetString('PyExc_TypeError', message)
 
+    #--------------------------------------------------------------------
+    #                   Convert functions
+    #--------------------------------------------------------------------
 
-    def _body_scalar(self, variable, collect_var, error_check = False, tmp_variable = None):
+    def generate_scalar_converter_function(self, used_names, variable, check_is_needed = True):
```

Comment:
Does this need to be in the codegen? Or can it be elsewhere e.g. in the ast or maybe better in a new utilities folder. It is not AST but it is not printing either, it is logic

Reply:
Then again, maybe that's true for most things in this file

Reply:
yes I'm going to move them all to the ast/x_wrapper once i finish